src/openrct2-ui/interface/InGameConsole.cpp

```diff
@@ -0,0 +1,313 @@
+#include <openrct2/interface/Window.h>
+#include <openrct2/localisation/Language.h>
+#include <openrct2/Version.h>
+#include "InGameConsole.h"
+
+using namespace OpenRCT2::Ui;
+
+static InGameConsole _inGameConsole;
+
+InGameConsole::InGameConsole()
+{
+    InteractiveConsole::WriteLine(OPENRCT2_NAME " " OPENRCT2_VERSION);
+    InteractiveConsole::WriteLine("Type 'help' for a list of available commands. Type 'hide' to hide the console.");
+    InteractiveConsole::WriteLine("");
+    WritePrompt();
+}
+
+void InGameConsole::WritePrompt()
+{
+    InteractiveConsole::WriteLine("> ");
+}
+
+void InGameConsole::Input(CONSOLE_INPUT input)
+{
+    switch (input) {
+    case CONSOLE_INPUT_LINE_CLEAR:
+        ClearInput();
+        RefreshCaret();
+        break;
+    case CONSOLE_INPUT_LINE_EXECUTE:
+        if (_consoleCurrentLine[0] != '\0') {
+            HistoryAdd(_consoleCurrentLine);
+            Execute(_consoleCurrentLine);
+            WritePrompt();
+            ClearInput();
+            RefreshCaret();
+        }
+        ScrollToEnd();
+        break;
+    case CONSOLE_INPUT_HISTORY_PREVIOUS:
+        if (_consoleHistoryIndex > 0) {
+            _consoleHistoryIndex--;
+            memcpy(_consoleCurrentLine, _consoleHistory[_consoleHistoryIndex], 256);
+        }
+        _consoleTextInputSession->Size = strlen(_consoleTextInputSession->Buffer);
+        _consoleTextInputSession->Length = utf8_length(_consoleTextInputSession->Buffer);
+        _consoleTextInputSession->SelectionStart = strlen(_consoleCurrentLine);
+        break;
+    case CONSOLE_INPUT_HISTORY_NEXT:
+        if (_consoleHistoryIndex < _consoleHistoryCount - 1) {
+            _consoleHistoryIndex++;
+            memcpy(_consoleCurrentLine, _consoleHistory[_consoleHistoryIndex], 256);
+            _consoleTextInputSession->Size = strlen(_consoleTextInputSession->Buffer);
+            _consoleTextInputSession->Length = utf8_length(_consoleTextInputSession->Buffer);
+            _consoleTextInputSession->SelectionStart = strlen(_consoleCurrentLine);
+        } else {
+            _consoleHistoryIndex = _consoleHistoryCount;
+            ClearInput();
+        }
+        break;
+    case CONSOLE_INPUT_SCROLL_PREVIOUS:
+    {
+        sint32 scrollAmt = GetNumVisibleLines() - 1;
+        Scroll(scrollAmt);
+        break;
+    }
+    case CONSOLE_INPUT_SCROLL_NEXT:
+    {
+        sint32 scrollAmt = GetNumVisibleLines() - 1;
+        Scroll(-scrollAmt);
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void InGameConsole::ClearInput()
+{
+    _consoleCurrentLine[0] = 0;
+    if (_isOpen) {
+        context_start_text_input(_consoleCurrentLine, sizeof(_consoleCurrentLine));
+    }
+}
+
+void InGameConsole::HistoryAdd(const utf8 * src)
+{
+    if (_consoleHistoryCount >= CONSOLE_HISTORY_SIZE) {
+        for (sint32 i = 0; i < _consoleHistoryCount - 1; i++)
+            memcpy(_consoleHistory[i], _consoleHistory[i + 1], CONSOLE_INPUT_SIZE);
```

Comment:
Happy to see the in-game console move to the UI subproject. However, this seems to be undoing the refactoring from https://github.com/OpenRCT2/OpenRCT2/pull/7064. Specifically, the console has been using a `deque` for its history since that PR.

Reply:
I haven't change anything about the way it works, this is a straight move. This code still exists in that PR.

Reply:
It appears you're right — sorry. I wonder why these instances of `memcpy` weren't replaced at the time, then… Others were replaced with `deque::push_back` and the likes.